### ISSUE TITLE
Issue with APA Formatting when no author is specified

### ISSUE THREAD

**Issue by QaidVoid** [typst/typst#2762]
### Description

When using references without authors, the generated output format is incorrect.

Example bib,
```bib
@misc{sdm,
  title = {Definition and objectives of systems development},
  url = {https://www.opentextbooks.org.hk/ditatopic/25323},
  language = {en},
  urldate = {2023-11-24},
  journal = {Open Textbooks for Hong Kong},
  date = {2016-01-19},
}
```

Expected Citation Output:
(_Definition and objectives of systems development_, 2016)

Actual Citation Output:
(2016)

Expected Bibliography Output:
Definition and Objectives of Systems Development. (2016, January 19). Open Textbooks for Hong Kong. https://www.opentextbooks.org.hk/ditatopic/25323

Actual Bibliography Output:
(2016, January 19). https://www.opentextbooks.org.hk/ditatopic/25323

### Operating system

Linux

### Typst version

- [X] I am using the latest version of Typst

**Comment by Leedehai** [comment]
TLDR: I can reproduce this issue.

Long version:

I added these stanzas to the end of assets/files/works.bib
```
@misc{book_has_author,
  title = {Book Title of Has Author},
  url = {https://www.opentextbooks.org.hk/ditatopic/25323},
  language = {en},
  urldate = {2023-11-24},
  journal = {Open Textbooks for Hong Kong},
  author = {Strong, Edward},
  date = {2016-01-19},
}

@misc{book_no_author,
  title = {Book Title of No Author},
  url = {https://www.opentextbooks.org.hk/ditatopic/25323},
  language = {en},
  urldate = {2023-11-24},
  journal = {Open Textbooks for Hong Kong},
  date = {2016-01-19},
}
```
and ran the test in tests/typ/meta/bibliography.typ with new test cases:
```
---
== Missing elements APA
See @book_has_author, @book_no_author.
#bibliography("/files/works.bib", style:"apa")
---
== Missing elements APS
See @book_has_author, @book_no_author.
#bibliography("/files/works.bib", style:"american-physics-society")
---
== Missing elements IEEE
See @book_has_author, @book_no_author.
#bibliography("/files/works.bib", style:"ieee")
```

The APA style reproduced this issue. I'm not sure of the APS style (may need to look up the APS style guide, but haven't got the time). The IEEE style can correctly display both items: one with and one without author.
<img width="125" alt="Screenshot 2023-11-27 at 18 07 27" src="https://github.com/typst/typst/assets/18319900/680778cb-ada7-47f9-8d72-430e511ba100">

Given that some style works and some doesn't, _it appears_ the bug lies in the CSL data that configures the citation/bibliography formats, not in Typst or Hayagriva code themselves.

(I'm not claiming this issue; others feel free to take a crack at it)

**Comment by zepinglee** [comment]
> Given that some style works and some doesn't, _it appears_ the bug lies in the CSL data that configures the citation/bibliography formats, not in Typst or Hayagriva code themselves.

I don't think the bug lies in the `.csl` style. The following are outputted from Zotero's style editor with imported .bib data as well as the APA CSL style. Both entries are as expexted.

<img width="628" alt="Screenshot 2023-11-28 at 10 17 03" src="https://github.com/typst/typst/assets/12290822/da088476-f97c-4902-a418-4ee36f2a94c1">

It looks to be a problem with the [`substitue`](https://docs.citationstyles.org/en/stable/specification.html#substitute) implementation.

**Comment by Leedehai** [comment]
Thanks! Given that it's not in components as high as Typst and Hayagriva, and not in components as low as CSL files, it seems to be in the intermediate [citationberg](https://github.com/typst/citationberg) @reknih ?

Specifically https://github.com/typst/citationberg/blob/main/src/lib.rs#L2174-L2175 mentioned [`cs:substitute`](https://docs.citationstyles.org/en/stable/specification.html#substitute).
```
/// Possible children for a `cs:names` element.
#[derive(Debug, Clone, Eq, PartialEq, Hash, Deserialize, Serialize)]
#[serde(rename_all = "kebab-case")]
pub enum NamesChild {
    /// A `cs:name` element.
    Name(Name),
    /// A `cs:et-al` element.
    EtAl(EtAl),
    /// A `cs:label` element.
    Label(VariablelessLabel),
    /// A `cs:substitute` element.
    Substitute(Substitute),
}
```
